### PR TITLE
Refactor - separate EPG

### PR DESCRIPTION
**Pull request by wedebe**
Separate EPG code

This change:
- is the start of EPG code refactoring & optimisation
- moves EPG code into its own controller

This change does not:
- optimise EPG code

Known issues:
- none known - `Modern`, `Classic` and direct endpoints tested

### CODE REVIEW

File: plugin/controllers/web.py [hunk 1]
```diff
@@ -1505,34 +1511,49 @@ def P_epgxmltv(self, request):
...
+		bqRef = getUrlArg(request, "bRef")
+		ret = getBouquetNowNextEpg(bqRef, 'nowNext', self.isJson)
 		info = getCurrentService(self.session)
-		ret = getBouquetNowNextEpg(getUrlArg(request, "bRef"), -1, self.isJson)
 		ret["info"] = info
 		return ret
 
-	def P_epgservicelistnownext(self, request):
-		res = self.testMandatoryArguments(request, ["sList"])
+	def P_epgmultichannelnownext(self, request):
```
Comment (jbleyel) [inline plugin/controllers/web.py:1546]: Don't change the api.

File: plugin/controllers/ajax.py [hunk 2]
```diff
@@ -95,12 +95,14 @@ def P_eventdescription(self, request):
...
 	def P_event(self, request):
 		event = getEvent(getUrlArg(request, "sref"), getUrlArg(request, "idev"))
-		event['event']['recording_margin_before'] = config.recording.margin_before.value
-		event['event']['recording_margin_after'] = config.recording.margin_after.value
-		event['at'] = HASAUTOTIMER
-		event['transcoding'] = TRANSCODING
-		event['moviedb'] = config.OpenWebif.webcache.moviedb.value if config.OpenWebif.webcache.moviedb.value else EXT_EVENT_INFO_SOURCE
-		event['extEventInfoProvider'] = extEventInfoProvider = getEventInfoProvider(event['moviedb'])
+		if event:
+			# TODO: this shouldn't really be part of an event's data
```
Comment (jbleyel) [inline plugin/controllers/ajax.py:99]: YES. it's .. otherwice you need to change the template

File: plugin/controllers/web.py (hunk 1 above)
Comment (wedebe) [reply plugin/controllers/web.py]: Reverted in #0e0c096

Comment (jbleyel) [reply plugin/controllers/web.py]: you can create an additional api call with the new name if you like

Comment (wedebe) [reply plugin/controllers/web.py]: > you can create an additional api call with the new name if you like

Will do, I'll map it to the existing endpoint. I plan to make all of the endpoints more consistent (eg. there's a ton of param name mismatches) while keeping backwards compatibility.

**Comment by jbleyel** [comment]
All your changes needs a very intensive test to prevent any api result differences.
We may can extend the testsuite.

https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/tree/master/testsuite

**Comment by jbleyel** [comment]
For now you can only do this:
* take 2 boxes one with the current code and one with the new code
* call each changed api and compare the results

This is a very complex task. ;-)

File: plugin/controllers/web.py [hunk 3]
```diff
@@ -1669,7 +1702,7 @@ def P_getcurrent(self, request):
 
 		"""
 		info = getCurrentService(self.session)
-		now = getNowNextEpg(info["ref"], 0, self.isJson)
+		now = getNowNextEpg(info["ref"], 'now', self.isJson)
```
Comment (jbleyel) [inline plugin/controllers/web.py:1705]: 'now' should be EPG.NOW as a constant of new EPG class.

File: plugin/controllers/web.py [hunk 4]
```diff
@@ -2180,7 +2220,14 @@ def P_loadepg(self, request):
 		Returns:
 			HTTP response with headers
 		"""
-		return loadEpg()
+		from Plugins.Extensions.OpenWebif.controllers.epg import Epg
+		Epg().load()
+
+		# TODO: get result from called function
```
Comment (jbleyel) [inline plugin/controllers/web.py:2226]: not that easy.

`void eEPGCache::load()
`

File: plugin/controllers/epg.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+# Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+##########################################################################
+
+from time import localtime, strftime, gmtime
+import json
+
+from enigma import eEPGCache, eServiceReference
+from Components.config import config
+
+DEBUG_ENABLED = True
```
Comment (jbleyel) [inline plugin/controllers/epg.py:30]: Is this only temporary?
If not then make an extra config setting for debug and move the debug function to the defaults controller.

File: plugin/controllers/epg.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+		debug(epgEvent)
+		epgEvent = epgEvent[0] if len(epgEvent) > 0 else None
+
+		# debug(json.dumps(epgEvent, indent = 2))
+		# debug(epgEvent)
+		return epgEvent
+
+
+	#TODO: investigate using `get event by time`
+	def getNowEvent(self, sRef):
```
Comment (jbleyel) [inline plugin/controllers/epg.py:239]: getNowEvent and getNextEvent is duplicate code

File: plugin/controllers/epg.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+
+
+	#TODO: investigate using `get event by time`
+	def getNextEvent(self, sRef):
+		debug("[[[   getNextEvent(%s)   ]]]" % (sRef))
+		# from Components.Sources.EventInfo import EventInfo
+		# debug(EventInfo(self.session.nav, EventInfo.NEXT).getEvent()) #.getBeginTime()
+
+		# sRef is not expected to be an instance of eServiceReference
+		criteria = ['TBDISEX', (sRef, MATCH_EVENT_AFTER_GIVEN_START_TIME, TIME_NOW)] # remove X
```
Comment (jbleyel) [inline plugin/controllers/epg.py:274]: TIME_NOW is wrong

File: plugin/controllers/epg.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+		criteria = ['IBDTSENCW', (sRef, MATCH_EVENT_INTERSECTING_GIVEN_START_TIME, startTime, endTime)]
+		epgEvents = self._gogogo(criteria)
+
+		# debug(json.dumps(epgEvents, indent = 2))
+		debug(epgEvents)
+		return epgEvents
+
+
+	#TODO: investigate using `get event by time`
+	def getChannelNowEvent(self, sRef):
```
Comment (jbleyel) [inline plugin/controllers/epg.py:297]: See getNowEvent/NextEvent

File: plugin/controllers/epg.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+		# sRef is not expected to be an instance of eServiceReference
+		epgEvents = self._gogogo(criteria)
+
+		# debug(json.dumps(epgEvents, indent = 2))
+		debug(epgEvents)
+		return epgEvents
+
+
+	#TODO: investigate using `get event by time`
+	def getBouquetNextEvents(self, sRefs):
```
Comment (jbleyel) [inline plugin/controllers/epg.py:390]: See getNow/NextEvent

File: plugin/controllers/models/services.py [hunk 10]
```diff
@@ -934,17 +923,16 @@ def getBouquetNowNextEpg(ref, servicetype, encode=False):
...
+	epg = Epg()
+
+	if servicetype == 'now': #TODO: use epg.NOW_E
+		events = epg.getBouquetNowEvents(sRefs)
+	elif servicetype == 'next': #TODO: use epg.NEXT_E
+		events = epg.getBouquetNextEvents(sRefs)
 	else:
-		for service in services.getContent('S'):
-			search.append((service, servicetype, -1))
+		events = epg.getBouquetNowNextEvents(sRefs)
```
Comment (jbleyel) [inline plugin/controllers/models/services.py:934]: you need only one function called 

epg.getBouquetEvents

with the parameter now, next and now_next

File: plugin/controllers/models/services.py [hunk 11]
```diff
@@ -1108,9 +1091,9 @@ def getMultiEpg(self, ref, begintime=-1, endtime=None, Mode=1):
 	# Check if an event has an associated timer. Unfortunately
 	# we cannot simply check against timer.eit, because a timer
 	# does not necessarily have one belonging to an epg event id.
-	def getTimerEventStatus(event, eventLookupTable):
+	def getTimerEventStatus(event, eventLookupTable): #TODO: this seems to be duplicated
 		startTime = event[eventLookupTable.index('B')]
-		endTime = event[eventLookupTable.index('B')] + event[eventLookupTable.index('D')] - 120
+		endTime = event[eventLookupTable.index('B')] + event[eventLookupTable.index('D')] - 120 # TODO: find out what this 120 means
```
Comment (jbleyel) [inline plugin/controllers/models/services.py:1096]: - 120

means 120 seconds in the past

Comment (jbleyel) [reply plugin/controllers/models/services.py:1096]: end = begin + duration - 120

Comment (jbleyel) [reply plugin/controllers/models/services.py:1096]: I'm not sure but I remember there was a timer overlapping issue.

**Comment by jbleyel** [comment]
By the way .. you're refactor is not that bad but you may hitting the bee hive. ;-)

**Comment by wedebe** [comment]
> By the way .. you're refactor is not that bad but you may hitting the bee hive. ;-)

Thanks, and oh I know it! 🙈 I've just picked up the stick so far... 😆

Would you rather I fix based on your comments and we merge what's done so far, or should I continue with changes and do one major merge (the latter would be more difficult to pinpoint any issues).

File: plugin/controllers/models/services.py (hunk 11 above)
Comment (wedebe) [reply plugin/controllers/models/services.py:1096]: I'd really like to change this to a named const but don't want to add confusion by using the wrong name.

**Comment by wedebe** [comment]
> * `Modern`, `Classic` and direct endpoints tested

Modern, Classic and direct endpoints tested

**Comment by wedebe** [comment]
> All your changes needs a very intensive test to prevent any api result differences. We may can extend the testsuite.
> 
> https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/tree/master/testsuite

I've manually tested Modern, Classic and direct endpoints.

**Comment by wedebe** [comment]
> For now you can only do this:
> 
> * take 2 boxes one with the current code and one with the new code
> * call each changed api and compare the results
> 
> This is a very complex task. ;-)

I know only too well 😄 - I had a whole set of browser tabs open to save responses from the development image then rebooted to a clean image to save and compare those responses.

File: plugin/controllers/web.py (hunk 4 above)
Comment (wedebe) [reply plugin/controllers/web.py:2226]: Ah, in that case maybe a `EPG load/save submitted` message?

File: plugin/controllers/epg.py (hunk 5 above)
Comment (wedebe) [reply plugin/controllers/epg.py]: I've set to `False`, will do separate work to move to utilities.py

File: plugin/controllers/epg.py (hunk 6 above)
Comment (wedebe) [reply plugin/controllers/epg.py:242]: There's a difference in event fields - `TBDCISEX` / `TBDISEX`; I mapped functions directly to start the refactor as I didn't want to rework the indices (I plan to name each field).

File: plugin/controllers/epg.py (hunk 8 above)
Comment (wedebe) [reply plugin/controllers/epg.py:300]: As comment above.

File: plugin/controllers/epg.py (hunk 9 above)
Comment (wedebe) [reply plugin/controllers/epg.py:393]: As comment above.

File: plugin/controllers/models/services.py (hunk 10 above)
Comment (wedebe) [reply plugin/controllers/models/services.py:934]: I actually want to split now, next and nowNext functions for simplicty, then internally call the same function.

File: plugin/controllers/epg.py (hunk 6 above)
Comment (wedebe) [reply plugin/controllers/epg.py]: getNowEvent and getNextEvent removed

File: plugin/controllers/epg.py [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+TIME_NOW = -1
+
+def debug(msg):
+	if DEBUG_ENABLED:
+		print(msg)
+
+class Epg():
+	NOW = 'now'
+	NEXT = 'next'
+	NOW_NEXT = 'nowNext'
```
Comment (jbleyel) [inline plugin/controllers/epg.py:49]: This can be integer

File: plugin/controllers/epg.py [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+		return epgEvents
+
+
+	def getBouquetEvents(self, sRefs, startTime, endTime=-1):
+		debug("[[[   getBouquetEvents(%s, %s, %s)   ]]]" % (sRefs, startTime, endTime))
+		# prevent crash #TODO: investigate if this is still needed (if so, use now + year or similar)
+		if endTime > 100000:
+			endTime = -1
+
+		criteria = ['IBDCTSERNWX'] # remove X
```
Comment (jbleyel) [inline plugin/controllers/epg.py:316]: yes .. NO X

File: plugin/controllers/epg.py [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+
+		# sRef is not expected to be an instance of eServiceReference
+		epgEvents = self._queryEPG(criteria)
+
+		# debug(json.dumps(epgEvents, indent = 2))
+		debug(epgEvents)
+		return epgEvents
+
+
+	def getBouquetNowNextEvents(self, sRefs):
```
Comment (jbleyel) [inline plugin/controllers/epg.py:361]: Now, Next and NowNext can be one function

Comment (wedebe) [reply plugin/controllers/epg.py:361]: I plan to have these three as a façade to one function, which will have the important code consolidated.

**Comment by jbleyel** [comment]
Are you finished with this or do you wanna do more tests?

**Comment by wedebe** [comment]
> Are you finished with this or do you wanna do more tests?

I have more to do on the EPG code separation but have done a round of testing successfully on what has been committed here; I'm happy to merge this as it is, as I don't know when I'll have time to keep going (what's here is pretty much a direct mapping of existing functions - next will be consolidation, optimisation and new functionality.

**Pull request merged**